OpenRA.Mods.Common/Traits/AffectsShroud.cs

```diff
@@ -86,30 +87,48 @@ PPos[] ProjectedCells(Actor self)
 				.ToArray();
 		}
 
-		void ITick.Tick(Actor self)
+		void INotifyVisualPositionChanged.VisualPositionChanged(Actor self, byte oldLayer, byte newLayer)
 		{
 			if (!self.IsInWorld)
 				return;
 
 			var centerPosition = self.CenterPosition;
 			var projectedPos = centerPosition - new WVec(0, centerPosition.Z, centerPosition.Z);
 			var projectedLocation = self.World.Map.CellContaining(projectedPos);
-			var traitDisabled = IsTraitDisabled;
-			var range = Range;
 			var pos = self.CenterPosition;
 
 			if (Info.MoveRecalculationThreshold.Length > 0 && (pos - cachedPos).LengthSquared > Info.MoveRecalculationThreshold.LengthSquared)
 				dirty = true;
 
-			if (!dirty && cachedLocation == projectedLocation && cachedRange == range && traitDisabled == CachedTraitDisabled)
+			if (!dirty && cachedLocation == projectedLocation)
 				return;
 
-			cachedRange = range;
 			cachedLocation = projectedLocation;
-			CachedTraitDisabled = traitDisabled;
 			cachedPos = pos;
 			dirty = false;
```

Comment:
I suspect (but haven't tested) that this flag needs to be checked and reset in `Tick` to avoid another regression - we won't be calling `VisualPositionChanged` again after resetting to `MovementType.None`, right?

Reply:
Also sorry, i'm not intentionally dragging these review comments out - i'm just noticing new things each time I look over this commit...

Reply:
Fixed by directly calling `UpdateShroudCells` in `MovementTypeChanged`.